CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/AnalyticViewTopComponent.java

```diff
@@ -109,16 +109,6 @@ public AnalyticViewTopComponent() {
                 prerequisiteAttributes.get(attribute).add(plugin);
             });
         });
-        prerequisiteAttributes.forEach((attribute, plugins) -> {
```

Comment:
Why have you removed all the attribute listeners? Also, is there a reason you haven't implemented the structure and state listeners above?

Reply:
Regarding the attribute listeners, 
They seemed to be running over most if not all plugin parameters and updating their parameters to the default parameters. Do you have a suggestion on how this would be possible to work around, whilst still allowing them to be added for the plugins not on this repo?
I can take a look at refactoring the code to include these state listeners if you feel they would benefit as opposed to the current solution.

Reply:
So what _should_ be happening here is that individual analytics can specify prerequisite attributes (ie. attributes they rely on for their parameters) using ```AnalyticPlugin.getPrerequisiteAttributes``` along with a way of handling changes to those attributes using ```AnalyticPlugin.onPrerequisiteAttributeChange```. This block of code is generically attaching listeners against all specified prerequisite attributes and handling them with the provided handler functions. I would expect many analytics to be affected by this as ```ScoreAnalyticPlugin``` uses this to listen for changes to the ```AnalyticConcept.TransactionAttribute.TYPE``` attribute in order to update its "Transaction Types" parameter when types are added to or removed from the graph. If it's not quite behaving on an update, then it might be worth taking a look at ```ScoreAnalyticPlugin.onPrerequisiteAttributeChange``` and ```FactAnalyticPlugin.onPrerequisiteAttributeChange```.

Reply:
Hi @cygnus-x-1 I've noticed that within the section I removed for listeners and prerequisite attributes, there is a call to createParameters().

https://github.com/constellation-app/constellation/blob/b9389f271320f33bb94ee8115ce96376d5e344da/CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/analytics/AnalyticPlugin.java#L80-L83

I believe that when it hits line 115, the plugins that do not implement any special behaviour just act out the superclass method above. Which just returns the default parameters from a plugin.
https://github.com/constellation-app/constellation/blob/b9389f271320f33bb94ee8115ce96376d5e344da/CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/AnalyticViewTopComponent.java#L112-L122

These are then set at line 117 which overrides the different parameter values at runtime. 

As createParameters() is utilized by many other classes, changing the default behaviour may come with unwanted bugs. How could I handle this to keep it from breaking other code.

I'll also push the rest of the code, this is the last of the pull request fixes currently in review.

Reply:
That sounds right - if an ```AnalyticPlugin``` does not have any parameters, then it'll simply use an empty ```PluginParameters```  inherited from the abstract base class. 

The listeners in the ```AnalyticViewTopComponent``` will then react to changes to a "prerequisite attribute" by updating that plugins parameters (which defaults to no parameters) using that plugins ```onPrerequisiteAttributeChange``` implementation (which defaults to a no-op). Note that this code also takes a copy of the parameters and updates ```SelectableAnalyticPlugin.updatedParameters``` instead of ```SelectableAnalyticPlugin.parameters``` in order to avoid events being triggered by changes to these parameters until the Analytic View is ready for them (```AnalyticConfigurationPane.updateSelectablePluginsParameters```).

```Plugin.createParameters``` is simply designed to set-up and return the default parameters for a plugin, generally to be used at the beginning of a plugin lifecycle. It is designed to be overridden to make it specific to a plugin implementation, and I believe that's exactly how it's being used here. Are you seeing any specific issues manifesting with this code?

Reply:
> Plugin.createParameters is simply designed to set-up and return the default parameters for a plugin, generally to be used at the beginning of a plugin lifecycle. It is designed to be overridden to make it specific to a plugin implementation, and I believe that's exactly how it's being used here.

You are not wrong in saying this, It currently gets the default `PluginParameters` during the `createParameters()` call. Then, obviously it `.copy()` 's the default parameters values.
The main issue I see is that within `.manualUpdate()`, all listeners are triggered to update things. Including the PrerequisiteAttributeListeners. The graph is read from and state attribute written to first. Then when you launch the Analytic View, it does a `.manualUpdate()`.

As you can see in this step through, the first image is line 117 getting the `SelectableAnalyticPlugin`. Note the red circle showing the boolean value read from the state attribute. 
![before updating parameters](https://user-images.githubusercontent.com/58677759/73501629-6accd700-4416-11ea-9a9a-e61a76a02f29.PNG)

In this second screenshot, It displays the value of the `updatedParameters` which are going to replace the current values. Note the highlighted boolean value down the bottom showing the `SelectableAnalyticPlugin` parameters value which will now be set to true after execution of this line. 
![image](https://user-images.githubusercontent.com/58677759/73501622-62749c00-4416-11ea-90e2-b09d09728fce.png)

All of this occurs when the following preconditions are met:
 - Graph is currently opened
 - Analytic View is being opened

Line 115 and 116 seem to work perfect for the use case when the prerequisite attribute changes, but on triggering of all handlers, it writes defaults to them which override the loaded parameters from state.

Reply:
Okay I think I understand the problem you're describing - essentially any parameters which are not updated by ```AnalyticPlugin.onPrerequisiteAttributeChange``` explicitly will instead be updated to the default value for that parameter when ```SelectableAnalyticPlugin.updatedParameters``` is copied across to ```SelectableAnalyticPlugin.parameters```, overwriting any other changes, such as those set by the state. Yeah that's a tough one - i'll have to think about that.